src/sparse_hessian.jl

```diff
@@ -1,3 +1,98 @@
+struct SparseADHessian <: ADNLPModels.ADBackend
+  d::BitVector
+  rowval::Vector{Int}
+  colptr::Vector{Int}
+  colors::Vector{Int}
+  ncolors::Int
+end
+
+function SparseADHessian(nvar, f, ncon, c!;
+  x0=rand(nvar),
+  alg::SparseDiffTools.SparseDiffToolsColoringAlgorithm = SparseDiffTools.GreedyD1Color(),
+  kwargs...,
+)
+  @variables xs[1:nvar]
+  xsi = Symbolics.scalarize(xs)
+  fun = f(xsi)
+  if ncon > 0
+    @variables ys[1:ncon]
+    ysi = Symbolics.scalarize(ys)
+    cx = similar(ysi)
+    fun = fun + dot(c!(cx,xsi), ysi)
+  end
+  S = Symbolics.hessian_sparsity(fun, ncon == 0 ? xsi : [xsi; ysi], full=false)
+  H = ncon == 0 ? S : S[1:nvar,1:nvar]
+  rows, cols, _ = findnz(H)
+  colors = matrix_colors(H, alg)
+  d = BitVector(undef, nvar)
+  ncolors = maximum(colors)
+  return SparseADHessian(d, H.rowval, H.colptr, colors, ncolors)
+end
+
+function get_nln_nnzh(b::SparseADHessian, nvar)
+  return length(b.rowval)
+end
+
+function hess_structure!(
+  b::SparseADHessian,
+  nlp::ADModel,
+  rows::AbstractVector{<:Integer},
+  cols::AbstractVector{<:Integer},
+)
+  rows .= b.rowval
+  for i = 1:(nlp.meta.nvar)
+    for j = b.colptr[i]:(b.colptr[i + 1] - 1)
+      cols[j] = i
+    end
+  end
+  return rows, cols
+end
+
+function sparse_hess_coord!(
+  ℓ::Function,
+  b::SparseADHessian,
+  x::AbstractVector,
+  vals::AbstractVector
+  )
+  nvar = length(x)
+  for icol = 1 : b.ncolors
+    b.d .= (b.colors .== icol)
+    res = ForwardDiff.derivative(t -> ForwardDiff.gradient(ℓ, x + t * b.d), 0)
+    for j = 1 : nvar
+      if b.colors[j] == icol
+        for k = b.colptr[j] : b.colptr[j+1] - 1
+          i = b.rowval[k]
+          vals[k] = res[i]
+        end
+      end
+    end
+  end
+  return vals
+end
+
+function hess_coord!(
+  b::SparseADHessian,
+  nlp::ADModel,
+  x::AbstractVector,
+  y::AbstractVector,
+  obj_weight::Real,
+  vals::AbstractVector,
+)
+  ℓ = get_lag(nlp, b, obj_weight, y)
+  sparse_hess_coord!(ℓ, b, x, vals)
+end
+
+function hess_coord!(
+  b::SparseADHessian,
+  nlp::ADModel,
+  x::AbstractVector,
+  obj_weight::Real,
+  vals::AbstractVector,
+)
+  ℓ(x) = obj_weight * nlp.f(x)
```

Comment:
```suggestion
  ℓ = get_lag(nlp, b, obj_weight)
```
as `nlp.f` is not defined for NLS